src/tracks/ui/BrushHandle.cpp

```diff
@@ -0,0 +1,567 @@
+   mMostRecentX = x;
+   mMostRecentY = y;
+
+   // Clip the coordinates
+   // TODO: Find ways to access the ClipParameters (for the mid)
+   int x1 = std::max(mRect.x + 10, std::min(x, mRect.x + mRect.width - 20));
+   int y1 = std::max(mRect.y + 10, std::min(y, mRect.y + mRect.height - 10));
+
+   mbCtrlDown = event.ControlDown();
+
+   if(!mpSpectralData->coordHistory.empty()){
+      int x0 = mpSpectralData->coordHistory.back().first;
+      int y0 = mpSpectralData->coordHistory.back().second;
+      int wd = mBrushRadius * 2;
+
+      int dx =  abs(x1-x0), sx = x0<x1 ? 1 : -1;
+      int dy = -abs(y1-y0), sy = y0<y1 ? 1 : -1;
+      int err = dx+dy, err2;
+
+      double posTime;
+      wxInt64 posFreq;
+
+      auto posToLongLong = [&](int x0){
+         posTime = viewInfo.PositionToTime(x0, mRect.x);
+         sampleCount sc = wt->TimeToLongSamples(posTime);
+         return sc.as_long_long();
+      };
+
+      // Line drawing (draw points until the start coordinate reaches the end)
+      while(true){
+         if (x0 == x1 && y0 == y1)
+            break;
+         err2 = err * 2;
+         if (err2 * 2 >= dy) { err += dy; x0 += sx; }
+         if (err2 * 2 <= dx) { err += dx; y0 += sy; }
+
+         // For each (x0, y0), draw circle based on Quadrant
+         int xm = x0, ym = y0;
+         int r = mBrushRadius;
+         int cx = -r, cy = 0, cerr = 2 - 2 * r;
+
+         if(mFreqSnapping){
+            // Correct the y coord (snap to highest energy freq. bin)
+            long long startSC = posToLongLong(xm);
+            wxInt64 targetFreq = PositionToFrequency(wt, 0, ym+cy, mRect.y, mRect.height);
+            if(auto *sView = dynamic_cast<SpectrumView*>(pView.get())){
+               wxInt64 resFreq = SpectralDataManager::FindFrequencySnappingBin(wt,
+                                                                               startSC,
+                                                                               mFreqSnappingRatio,
+                                                                               targetFreq);
+               ym = FrequencyToPosition(wt, resFreq, mRect.y, mRect.height);
+            }
+         }
+
+         do {
+            posFreq = PositionToFrequency(wt, 0, ym+cy, mRect.y, mRect.height);
+            HandleTimeFreqData(posToLongLong(xm-cx), posFreq);
+
+            posFreq = PositionToFrequency(wt, 0, ym-cx, mRect.y, mRect.height);
+            HandleTimeFreqData(posToLongLong(xm-cy), posFreq);
+
+            posFreq = PositionToFrequency(wt, 0, ym-cy, mRect.y, mRect.height);
+            HandleTimeFreqData(posToLongLong(xm+cx), posFreq);
+
+            posFreq = PositionToFrequency(wt, 0, ym+cx, mRect.y, mRect.height);
+            HandleTimeFreqData(posToLongLong(xm+cy), posFreq);
+
+            r = cerr;
+            if (r <= cy)
+               cerr += ++cy * 2 + 1;
+            if (r > cx || cerr > cy)
+               cerr += ++cx * 2 + 1;
```

Comment:
This is an algorithm for plotting an unfillled circle.  But shouldn't there be a filled circle?

The numerical tricks used for fast plotting may be less important here than correct calculation of the time and frequency bins that must be added.

Something like: